components/service/glean/src/main/java/mozilla/components/service/glean/LabeledMetricType.kt

```diff
@@ -0,0 +1,130 @@
+    override val sendInPings: List<String>,
+    val subMetric: T,
+    val labels: Set<String>? = null
+) : CommonMetricData {
+
+    override val defaultStorageDestinations: List<String> = (
+        subMetric as CommonMetricData).defaultStorageDestinations
+
+    private val logger = Logger("glean/LabeledMetricType")
+
+    companion object {
+        private const val MAX_LABELS = 16
+        private const val OTHER_LABEL = "__other__"
+        private val labelRegex = Regex("^[a-z_][a-z0-9_]{0,29}$")
+    }
+
+    private val seenLabels: MutableSet<String> = mutableSetOf()
+
+    /**
+     * Handles the label in the case where labels are predefined.
+     *
+     * If the given label is not in the predefined set of labels, returns [OTHER_LABEL], otherwise
+     * returns the label verbatim.
+     *
+     * @param label The label, as specified by the user
+     * @param adjusted label, possibly set to [OTHER_LABEL]
+     */
+    private fun getFinalStaticLabel(label: String): String {
+        return if (labels!!.contains(label)) label else OTHER_LABEL
+    }
+
+    /**
+     * Handles the label in the case where labels aren't predefined.
+     *
+     * If we've already seen more than [MAX_LABELS] unique labels, returns [OTHER_LABEL].
+     *
+     * Also validates any unseen labels to make sure they are snake_case and under 30 characters.
+     * If not, returns [OTHER_LABEL].
+     *
+     * @param label The label, as specified by the user
+     * @param adjusted label, possibly set to [OTHER_LABEL]
+     */
+    private fun getFinalDynamicLabel(label: String): String {
+        if (!seenLabels.contains(label)) {
+            if (seenLabels.size >= MAX_LABELS) {
+                return OTHER_LABEL
+            } else {
+                // Labels must be snake_case.
+                if (!labelRegex.matches(label)) {
+                    logger.error("Labels must be snake_case and < 30 characters. Got '$label'")
+                    return OTHER_LABEL
+                }
+                seenLabels.add(label)
+            }
+        }
+        return label
+    }
+
+    /**
+     * Get a copy of the subMetric with the name changed to the given `newName`.
+     *
+     * @param newName The new name for the metric.
+     * @returns A copy of subMetric with the new name.
+     */
+    private fun getMetricWithNewName(newName: String): T {
+        // Every metric that supports labels needs an entry here
+        return when (subMetric) {
+            is BooleanMetricType -> subMetric.copy(name = newName) as T
+            is CounterMetricType -> subMetric.copy(name = newName) as T
+            is StringListMetricType -> subMetric.copy(name = newName) as T
+            is StringMetricType -> subMetric.copy(name = newName) as T
+            is TimespanMetricType -> subMetric.copy(name = newName) as T
+            is UuidMetricType -> subMetric.copy(name = newName) as T
+            else -> throw IllegalStateException()
```

Comment:
nit: since this is throwing, should we add an error message to `IllegalStateException` so that we know what's happening while we panic ? :P